elasticgraph-warehouse/spec/integration/elastic_graph/warehouse/schema_definition/rake_tasks_spec.rb

```diff
@@ -0,0 +1,219 @@
+
+            warehouse_config = YAML.safe_load(read_artifact(Warehouse::DATA_WAREHOUSE_FILE))
+            expect(warehouse_config.keys).to eq(["tables"])
+            expect(warehouse_config["tables"].keys).to eq(["products"])
+            expect(warehouse_config["tables"]["products"]).to have_key("table_schema")
+            expect(warehouse_config["tables"]["products"]["table_schema"]).to start_with("CREATE TABLE IF NOT EXISTS products")
+          end
+
+          it "does not dump warehouse artifact when no warehouse tables are defined" do
+            write_warehouse_schema(table_defs: <<~EOS)
+              s.object_type "User" do |t|
+                t.field "id", "ID"
+                t.field "name", "String"
+                # No warehouse_table definition
+              end
+            EOS
+
+            output = run_rake_with_warehouse("schema_artifacts:dump")
+
+            expect(output.lines).not_to include(
+              a_string_including("Dumped", Warehouse::DATA_WAREHOUSE_FILE)
+            )
+            expect(read_artifact(Warehouse::DATA_WAREHOUSE_FILE)).to be_falsy
+          end
+
+          it "idempotently dumps warehouse artifacts" do
+            write_warehouse_schema(table_defs: <<~EOS)
+              s.object_type "Product" do |t|
+                t.field "id", "ID"
+                t.index "products" do |i|
+                  i.warehouse_table "products"
+                end
+              end
+            EOS
+
+            run_rake_with_warehouse("schema_artifacts:dump")
+
+            expect {
+              output = run_rake_with_warehouse("schema_artifacts:dump")
+              expect(output.lines).to include(a_string_including("already up to date", Warehouse::DATA_WAREHOUSE_FILE))
+            }.to maintain { read_artifact(Warehouse::DATA_WAREHOUSE_FILE) }
+          end
+
+          it "updates warehouse artifact when schema changes" do
+            write_warehouse_schema(table_defs: <<~EOS)
+              s.object_type "Product" do |t|
+                t.field "id", "ID"
+                t.index "products" do |i|
+                  i.warehouse_table "products"
+                end
+              end
+            EOS
+
+            run_rake_with_warehouse("schema_artifacts:dump")
+            original_content = read_artifact(Warehouse::DATA_WAREHOUSE_FILE)
+
+            write_warehouse_schema(table_defs: <<~EOS)
+              s.object_type "Product" do |t|
+                t.field "id", "ID"
+                t.field "name", "String"
+                t.index "products" do |i|
+                  i.warehouse_table "products"
+                end
+              end
+            EOS
+
+            expect {
+              run_rake_with_warehouse("schema_artifacts:dump", enforce_json_schema_version: false)
+            }.to change { read_artifact(Warehouse::DATA_WAREHOUSE_FILE) }
+              .from(original_content)
+              .to(a_string_including("products:", "name"))
```

Comment:
It seems like this assumes that `original` content does _not_ have `products:`, and `name` -- if it did, then the dumped artifact isn't changing to include them.

Should we assert that `original_content` lacks these substrings? Or if it includes them (I'm guessing it includes `products:`...) should we use different substrings here that `original_content` doesn't include?

One thought: `name` is such a generic term it could theoretically be a field name in the YAML data structure itself, apart from its usage as a column name. Maybe it's better to use a more unique field name such as `t.field "added_field"`, and then you can verify that `original_content` lacks `added_field` while the new artifact includes it?